docs/proposals/cron-wf-improvement-proposal.md

```diff
@@ -0,0 +1,102 @@
+# Proposal for Cron Workflows improvements
+
+## Introduction
+
+Currently, CronWorkflows are a great resource if we want to run recurring tasks to infinity. However, it is missing the ability to customize it, for example define how many times a workflow should run or how to handle multiple failures. I believe argo workflows would benefit of having more configuration options for cron workflows, to allow to change its behavior based on the result of its child’s success or failures. Below I present my thoughts on how we could improve them, but also some questions and concerns on how to properly do it.
+
+## Proposal
+
+This proposal discusses the viability of adding 2 more fields into the cron workflow configuration:
+
+```
+RunStrategy:
+ maxSuccess:
+ maxFailures:
+```
+
+`maxSuccess` - defines how many child workflows must have success before suspending the workflow schedule
+
+`maxFailures` - defines how many child workflows must fail before suspending the workflow scheduling. This may contain `Failed` workflows, `Errored` workflows or spec errors.
+
+For example, if we want to run a workflow just once, we could just set:
+
+```
+RunStrategy:
+ maxSuccess: 1
+```
+
+This configuration will make sure the controller will keep scheduling workflows until one of them finishes with success.
+
+As another example, if we want to stop scheduling workflows when they keep failing, we could configure the CronWorkflow with:
+
+```
+RunStrategy:
+ maxFailures: 2
+```
+
+This config will stop scheduling workflows if fails twice.
+
+## Total vs consecutive
+
+One aspect that needs to be discussed is whether these configurations apply to the entire life of a cron Workflow or just in consecutive schedules. For example, if we configure a workflow to stop scheduling after 2 failures, I think it makes sense to have this applied when it fails twice consecutively. Otherwise, we can have 2 outages in different periods which will suspend the workflow. On the other hand, when configuring a workflow to run twice with success, it would make more sense to have it execute with success regardless of whether it is a consecutive success or not. If we have an outage after the first workflow succeeds, which translates into failed workflows, it should need to execute with success only once. So I think it would make sense to have:
+
+- maxFailures - maximum number of **consecutive failures** before stopping the scheduling of a workflow
```

Comment:
I would propose adding another field to toggle between consecutive or total failures. It could be a boolean with default as `false`. 

Maybe: `consecFailures` or `consecFail`

This would make the default logic align for both `maxSuccess` and `maxFailures` which is simpler logic for a new user to understand. If they want consecutive failures counted, then they can set `consecFailures` to `true`.

Reply:
Maybe we could have the run strategy to be configurable by success or failure. For example:

```yaml
RunStrategy:
  onSuccess:
    maxSuccess: 3 
  onFail:
    maxFail:  3
    consecFail: false 
```

this would allow us to further upgrade the configurable behavior in the future without breaking the schema. For example,  if we decide to ship the back-off in another PR, you could just upgrade it to:

```yaml
RunStrategy:
  onSuccess:
    maxSucess: 3 
  onFail:
    maxFail:  3
    consecFail: true
    back-off:
      enabled: true
      factor: 2
```

However, I do agree with you, this may add a lot of complexity, and its best to keep it simple for now.

Reply:
I like this suggestion of yours:

> Maybe we could have the run strategy to be configurable by success or failure. this would allow us to further upgrade the configurable behavior in the future without breaking the schema. For example, if we decide to ship the back-off in another PR, you could just upgrade it

I'm up for this if you think it'll be simple enough to implement on the first attempt.